---
title: "cdQuPath Vignette: overview"
author: "Jianhong Ou"
date: "`r BiocStyle::doc_date()`"
package: "`r BiocStyle::pkg_ver('cdQuPath')`"
vignette: >
  %\VignetteIndexEntry{cdQuPath Vignette: overview}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output:
  BiocStyle::html_document:
    toc: true
    toc_float: true
    toc_depth: 4
    fig_caption: true
---

```{r, echo=FALSE, results="hide", warning=FALSE}
suppressPackageStartupMessages({
  library(cdQuPath)
  library(Seurat)
  library(dittoSeq)
  library(ggplot2)
  library(hoodscanR)
  library(SpatialExperiment)
  library(dplyr)
})
knitr::opts_chunk$set(warning=FALSE, message=FALSE)
```

# Introduction

This package presents a fundamental workflow for spatial phenotyping, cell type annotation, multi-nucleus detection, quality control, and calculating cell neighborhoods from whole slide images of ultra-high multiplexed tissue captured by 'Akoya Biosciences' assays. The workflow leverages QuPath, StarDist, Cellpose, Seurat, and hoodscanR.

The upstream code are available in the folder of `extdata`.
The detailed documentation will be provide in following sections.

The _cdQuPath_ aims to provide a easy to use tool for downstream analysis
with `Seurat` and `hoodscanR` for
quality control and neighborhoods analysis.

# Quick start

Here is an example using _cdQuPath_.

## Installation

First, please download and install [QuPath](https://qupath.github.io/) (v0.4.0 or newer). The cell segmentation requires the [cellpose](https://github.com/BIOP/qupath-extension-cellpose) and [stardist](https://github.com/qupath/qupath-extension-stardist) extension for `QuPath`. Don't forget to set the Python path for the `Cellpose` extension.

Then, download a copy of the Groovy script to your local machine. There are two scripts available: `cellpose_stardist_multiNucleis_underGUI.groovy` and `ExportCellDetectionMeasurement.groovy`.

```{sh, eval=FALSE}
git clone https://github.com/jianhong/cdQuPath.git
cd cdQuPath/inst/extdata/scripts
ls
```

Then, install _cdQuPath_ and other packages required to run 
the examples.

```{r installation, eval=FALSE}
library(BiocManager)
BiocManager::install("jianhong/cdQuPath")
```

## Step 1. Cell segmentation

This step involves merging the results from 
[Cellpose](https://www.cellpose.org/) and
[Stardist](https://github.com/stardist/stardist) analyses.

Navigate to line 25 of the `cellpose_stardist_multiNucleis_underGUI.groovy`
script and update the path of the `stardistModel` variable to match the 
file path of your `cell_seg_model.pb`.
Adjust the parameters from [line 26 to line 38](https://github.com/jianhong/codex_qupath/blob/main/scr/cellpose_stardist_multiNucleis_underGUI.groovy#L26-L38) accordingly.
You may want to change the `cytoChannels` and `measurementPercentileCutoff` according to your observations.
The `distanceCutoff` is the maximal cutoff distance of nucleus for multiple nucleus cells.

Then, select a region in the opened TIFF view window in QuPath. Start with a small region when testing the script, and expand it once you achieve the desired results. Please note that if you observe regional biases, you may want to exclude the biases regions. If it is un-avoidable, you may want to normalize the signals. We will discuss this later.

Finally, run the `cellpose_stardist_multiNucleis_underGUI.groovy` script by clicking the Run button in the Script Editor.

[![Youtube Video](https://img.youtube.com/vi/3SKKSDhlEkw/0.jpg)](https://youtu.be/3SKKSDhlEkw)

## Step 2. Export the cell segmentation

The script `ExportCellDetectionMeasurement.groovy` is designed to export cell segmentation data. Upon export, two files are generated and saved in a folder prefixed with `measurementsExport`. 

- The `.tsv` file is compatible with `createSeuratObj.R`.
- The `.geojson` file is intended for reloading into `QuPath`. 

These exported files contain comprehensive information, including cell area, locations, marker signal statistics, and nucleus classification.
The cell locations are particularly useful for neighborhood analysis.

## Step 3. Work in R

### Load library

```{r load_library}
library(dplyr)
library(cdQuPath)
library(Seurat)
library(ggplot2)
library(dittoSeq)
library(hoodscanR)
library(SpatialExperiment)
```

### Create a Seurat object from output `tsv` file.

```{r quick_start}
tsv <- system.file("extdata", "sample.qptiff.tsv.gz",
                   package = "cdQuPath",
                   mustWork = TRUE)
seu <- createSeuratObj(tsv, useValue = 'Median',
                       markerLocations = CodexPredefined$markerLocations)
seu
```

### Quality Control

The conventional method for evaluating dynamic range involves computing a signal-to-background (SNR) ratio. This is done by dividing the average intensity of the top 20 brightest cells by the average intensity of the weakest 10% of cells. An SNR value of 10 or higher is indicative of reliable image analysis. It is recommended that the SNR falls within the range of (10, Inf], typically exceeding 100. An SNR below 3 suggests that the antibodies are performing poorly.

```{r dynamicRanges}
## please make sure the default assay is 'RNA'
DefaultAssay(seu)
dr <- dynamicRanges(seu)
dr
## markers may not work;
## Please note that DAPI.01 is used for cell detection.
avoid_markers <- 
  names(dr$dynamic_range)[dr$dynamic_range<3 | is.na(dr$dynamic_range)]
avoid_markers
```

Verify the cell area and nucleus area for multi-nuclei.

```{r vlnplot1}
# Visualize QC metrics as a violin plot
colnames(seu[[]])
VlnPlot(seu,
        features = c("Cell.Area.um.2",
                     "Nucleus.Area.um.2",
                     'Nucleus.Cell.area.ratio'),
        split.by = 'Cell.isMultiNucleis',
        pt.size = ifelse(ncol(seu)>500, 0, 1),
        ncol = 3)
## if you want to save the plot, try
## ggsave(filename)
```
```{r vlnPlot2}
VlnPlot(seu, features = c('cd68', 'CD11c', 'cd34'),
        pt.size = ifelse(ncol(seu)>500, 0, 1),
        ncol = 3)
```


Now, we can proceed with filtering the cells. Upon observation, we notice that the average area of single-nucleus cells is below 500. Therefore, we establish the cutoff for cell area at approximately three times the size of single-nucleus cells.
Additionally, if the nucleus-to-cell area ratio is excessively high, it suggests potential issues with the cell area detection process. We will filter out those cells.

```{r filter}
keep <- seu$Cell.Area.um.2<1500 & seu$Nucleus.Cell.area.ratio < .9
seu <- subset(seu,
              cells = colnames(seu)[keep],
              features = names(dr$dynamic_range)[!is.na(dr$dynamic_range)])
seu
```

### Normalizing the data

The initial query is whether we are observing abnormally high or low intensity regions in specific sub-regions. If these variations are deemed unrelated to tissue-specific characteristics, it might be advisable to normalize each marker channel accordingly. The positional data of cells is contained within the reduction matrix labeled 'pos'.

```{r regionCheck}
nrow(seu)
suppressMessages(
  FeaturePlot(seu, slot='count',
            features = rownames(seu)[1],
            reduction = 'pos') + scale_y_reverse()
)
```

You may attempt to normalize the irregular sub-regions using LOESS (Local Polynomial Regression Fitting) with the `loessSmooth` function.

```{r loessSmooth, eval=FALSE, echo=FALSE, results="hide", warning=FALSE}
## this script will not run in the vignette.
## it is time consuming.
## under-development, do not try
seu_norm <- loessSmooth(seu)
FeaturePlot(seu_norm, slot='count',
            features = rownames(seu)[1],
            reduction = 'pos',
            ncol = 2) + scale_y_reverse()
```

The next step involves cell-wise normalization. The available normalization methods are 'LogNormalize', 'CLR' (centered log ratio), 'RC' (CPM), and 'zscore', which can be applied using the `normData` function. The normalization method except for 'zscore' is borrowed from `Seurat` package.

```{r normalization}
seu <- normData(seu, normalizationMethod='zscore')
```

### Identification of highly variable markers (marker selection)

In the background, this step will execute the `FindVariableFeatures` function and exclude the 'avoidMarkers' from the list of variable markers. Please ensure that 'nFeatures' is smaller than the total number of available markers.

```{r variableFeatures}
seu <- findVarMarkers(
    seu,
    avoidMarkers=c('DAPI.01', 'hist3h2a'),
    nFeatures = 10L)
VariableFeatures(seu)
plot1 <- VariableFeaturePlot(seu)
LabelPoints(plot = plot1, points = VariableFeatures(seu), repel = TRUE)
```

### Scaling the data and do cluster analysis

The following steps outline the standard process of single-cell analysis in the `Seurat` package. You may adjust the process accordingly to suit your needs.

```{r}
all.markers <- rownames(seu)
seu <- ScaleData(seu, features = all.markers)
seu <- RunPCA(seu, features = VariableFeatures(seu))
dim <- seq.int(ncol(Embeddings(seu$pca)))
VizDimLoadings(seu, dims = dim,
               nfeatures=length(VariableFeatures(seu)),
               reduction = "pca")
seu <- FindNeighbors(seu, dims = dim)
seu <- FindClusters(seu, resolution = 0.9)
seu <- RunUMAP(seu, reduction = "pca", dims = dim)
pt.size <- ifelse(ncol(seu)>500, 0, 2)
DimPlot(seu,
        reduction = 'umap', label = TRUE,
        pt.size = pt.size) + scale_y_reverse()
## check the distribution of the clusters in spatial position.
ncol <- ceiling(sqrt(length(unique(seu$seurat_clusters))))
DimPlot(seu,
        reduction = 'pos', label = FALSE,
        split.by = 'seurat_clusters',
        ncol = ncol, pt.size = pt.size) + scale_y_reverse()
## check the nucleis distribution
dittoBarPlot(seu,
             var = 'Cell.isMultiNucleis',
             group.by='seurat_clusters')
DimPlot(seu, reduction = 'pos', label = TRUE,
        split.by = 'Cell.isMultiNucleis',
        pt.size = pt.size) + scale_y_reverse()
VlnPlot(seu,
        features = c('Nucleus.Area.um.2',
                     'Cell.Area.um.2',
                     'Nucleus.Cell.area.ratio'),
        pt.size = pt.size)
## Dot plot
markers <- unique(FindAllMarkers(seu)$gene)
DotPlot(seu, features = markers, scale = FALSE) + coord_flip()
```

### Rescale data using Gaussian Mixture Model

In the downstream analysis, cell types are detected by fitting a multiple-class Gaussian Mixture Model (GMM) to the re-scaled data. The raw intensities are converted to probabilities, and a cutoff of 0.5 is set for cell type assignment.

Please run `fitGMM` for each sub-region of the image if the background of the sub-regions are not identical.

```{r gmm}
library(future.apply)
plan(multisession) # for multiple process.
set.seed(123) ## for model fit
seu <- fitGMM(seu)
DotPlot(seu, features = markers, assay = 'GMM', scale = FALSE) + coord_flip()
```


### Correlation heatmap

The correlation heatmap will display the signal similarity among different markers. The distance will be calculated using the `stats::dist` function. Users can experiment with different inputs such as z-scales, GMM data, and so on.

```{r cor}
DefaultAssay(seu)
p <- markerCorrelation(seu)
names(p)
## plot the correlations for data, here it is Z-scale data
p[['data']]
## use GMM rescaled data to plot the correlation
DefaultAssay(seu) <- 'GMM'
p_GMM <- markerCorrelation(seu)
p_GMM[['data']]
## 
DefaultAssay(seu) <- 'RNA'
```

### Assign cell types

The phenotype will utilize the re-scaled data obtained through GMM. Two methods are available for assigning cell types: 'Rank' and 'Boolean'. With the 'Rank' option, cell types are assigned based on the rank of probabilities for the positive markers. On the other hand, the 'Boolean' option assigns cell types according to the order of the classifier for the positive markers.

The metadata 'celltype' stores the assigned cell types for each Seurat cluster. Additionally, the metadata 'celltype_all' contains all possible cell types for each cell, determined by a cutoff of 0.5. Finally, the metadata 'celltype_first' represents the initial assignment.

```{r celltype}
seu <- detectCellType(seu)
table(seu$celltype)
## check the cell type annotation by clusters
DimPlot(seu, group.by = 'celltype')
DimPlot(seu, group.by = 'celltype', reduction = 'pos') + scale_y_reverse()
dittoBarPlot(seu,
             var = 'celltype',
             group.by='Cell.isMultiNucleis')
## check the first annotation of each cell, not restricted by seurat_cluster
DimPlot(seu, group.by = 'celltype_first', reduction = 'pos') + scale_y_reverse()
dittoBarPlot(seu,
             var = 'celltype_first',
             group.by='Cell.isMultiNucleis')
## check the signals for each celltype
Idents(seu) <- 'celltype'
DotPlot(seu, features = rownames(seu), scale = FALSE) + coord_flip() + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
## export the celltype
ct <- data.frame(Object.ID=colnames(seu),
                 x=seu$Cell.X,
                 y=seu$Cell.Y,
                 CellType=seu$celltype)
head(ct, n=2)
## if you see postfix like _1, then run
ct$Object.ID <- sub('_\\d+$', '', ct$Object.ID)
## output to same folder: celltype_csv <- sub('qptiff.tsv', 'celltype.csv', tsv)
celltype_csv <- tempfile(fileext = '.csv')
write.csv(ct, file = celltype_csv, row.names = FALSE)
## after exporting, you can load the celltypes back to QuPath by running the script load_celltype.groovy in the scripts folder.
```
[![Youtube Video](https://img.youtube.com/vi/OH8TKCxZKkM/0.jpg)](https://youtu.be/OH8TKCxZKkM)

### Merge multiple Seurat objects if necessary

If you have datasets from multiple sub-regions, you can process each one individually as shown above and then merge them. Occasionally, some images may not have the same background across different regions. To address this, you can detect the cells in each sub-region with a similar background or baseline and then perform GMM rescaling for each sub-region. Once this is done, you can combine the analyzed results to create a merged dataset.

```{r runSample2}
tsv2 <- system.file("extdata", "sample2.qptiff.tsv.gz",
                   package = "cdQuPath",
                   mustWork = TRUE)
seu2 <- createSeuratObj(tsv2, useValue = 'Median',
                       markerLocations = CodexPredefined$markerLocations)
seu2
keep <- seu2$Cell.Area.um.2<1500 & seu2$Nucleus.Cell.area.ratio < .9
table(keep)
seu2 <- subset(seu2,
              cells = colnames(seu2)[keep])
seu2 <- normData(seu2, normalizationMethod='zscore')
seu2 <- findVarMarkers(
    seu2,
    avoidMarkers=c('DAPI.01', 'hist3h2a'),
    nFeatures = 10L)
seu2 <- ScaleData(seu2, features = rownames(seu2))
seu2 <- RunPCA(seu2, features = VariableFeatures( seu2))
dim <- seq.int(ncol(Embeddings(seu2$pca)))
seu2 <- FindNeighbors(seu2, dims = dim)
seu2 <- FindClusters(seu2, resolution = 0.9)

seu2 <- fitGMM(seu2)
seu2 <- detectCellType(seu2)
seu2$pca <- NULL # to advoid different pca columns error.
seu$orig.ident <- 'region1'
seu2$orig.ident <- 'region2'
mergedSeu <- merge(x=seu, y=list(seu2), merge.dr = TRUE)
mergedSeu
Assays(mergedSeu)
p <- dittoBarPlot(mergedSeu,
             var = 'celltype_first',
             group.by='orig.ident')
p
## pie chart
plotData <- p$data %>%
  dplyr::arrange(desc(label)) %>%
  dplyr::group_by(grouping) %>%
  dplyr::mutate(y_pos = cumsum(percent) - 0.5*percent)
ggplot(plotData, aes(x="", y = percent, fill = label)) +
  geom_bar(stat="identity", width = 1) +
  geom_text(aes(x=1.7, y=y_pos,
                label=paste0(round(percent*100, digits = 1), '%')),
            color = 'black') +
  coord_polar(theta = 'y') +
  facet_wrap(~grouping) +
  theme_void() +
  scale_fill_manual(values = dittoColors())

FeaturePlot(mergedSeu, slot='count',
            features = rownames(mergedSeu)[1],
            reduction = 'pos',
            ncol = 2) + scale_y_reverse()
DimPlot(mergedSeu, group.by = 'celltype', reduction = 'pos') + scale_y_reverse()
## use GMM rescaled data to plot the correlation
DefaultAssay(mergedSeu) <- 'GMM'
p_GMM <- markerCorrelation(mergedSeu, layers = 'data')
p_GMM[['data']]
## 
DefaultAssay(mergedSeu) <- 'RNA'
```

### Detect the celltype by classifier

The simplest classifier is a list of markers.

```{r}
CodexPredefined$classifier
```

You can also try to detect the cell type by multiple markers with weight for 
each celltype.

```{r}
CodexPredefined$classifier2
mergedSeu <- detectCellType(mergedSeu,
                            classifier = CodexPredefined$classifier2,
                            celltypeColumnName = 'celltype2')
dittoBarPlot(mergedSeu,
             var = 'celltype2',
             group.by='Cell.isMultiNucleis')
```


You can also try to detect the cell type by a generalized k-Nearest Neighbors Classifier.

```{r celltypeDetection2}
classifier <- readRDS(system.file('extdata', 'classifier.CSA.res0.3.rds', 
                      package='cdQuPath'))
## this request lots of memory.
mergedSeu <- detectCellType(mergedSeu,
               classifier=classifier,
               celltypeColumnName='celltypeByKNN')
table(mergedSeu$celltypeByKNN)
dittoBarPlot(mergedSeu,
             var = 'celltypeByKNN',
             group.by='orig.ident')
```

### Align the single cell RNAseq data to Codex data by KNN classifier

Below is sample code to create a classifier for your data and align single-cell RNA-seq data with Codex data.

Codex data has a limited number of cell markers compared to the 10,000+ features detected by single-cell RNA-seq. To align the scRNA-seq matrix with the Codex matrix, it is crucial to minimize the effects of sparsity, noise, and collinearity among individual cells. This is achieved by performing distance calculations using pseudobulk analysis, which aggregates gene expression from individual cells into group-level values. By treating each group as a single cell, this approach reduces heterogeneity within cell clusters.

The resulting alignment matrix enables the prediction of transcriptomes using Codex spatial information.

```{r alignCells}
## read the protein-gene map from the Codex (protein level) to
## transcriptome (gene level)
markers <- readRDS(system.file('extdata', 'markers.name.map.rds',
                                package='cdQuPath'))
tail(markers)
## load the scRNA-seq data
# scRNAseqSeu <- readRDS('int.rds')
## do cluster, set higher resolution to get more class
# scRNAseqSeu <- FindNeighbors(scRNAseqSeu, dims = 1:20)
## set a big resolution to split the cells into a big number of clusters.
# scRNAseqSeu <- FindClusters(scRNAseqSeu, resolution = 100)
## create a pseudobulking for each cluster
# pseudo <- AggregateExpression(scRNAseqSeu, assays = "RNA",
#                               return.seurat = TRUE,
#                               group.by = c("seurat_clusters"))
pseudo <- readRDS(system.file('extdata', 'pseudobulk.scRNAseq.rds',
                              package='cdQuPath'))
library(future.apply)
plan(multisession)
## will cost several mins
classifier <- buildClassifierFromSeurat(pseudo,
                                        markerList = markers)
alignedCodex <- alignScRNAseqToCodex(codexSeu = mergedSeu,
                                      scSeu = pseudo,
                                      classifier = classifier)
## plot any gene expression with DarkTheme()
FeaturePlot(alignedCodex, slot='data',
            features = rownames(pseudo)[1],
            reduction = 'pos',
            ncol = 2) + scale_y_reverse() + DarkTheme()
```

### Detect the celltype by other tools

Here we will show the sample code to detect the cell types by [CELESTA](https://github.com/plevritis-lab/CELESTA).

```{r CELESTA, eval=FALSE}
if (!require("CELESTA", quietly = TRUE)){
  BiocManager::install("plevritis-lab/CELESTA")
}
library(CELESTA)
library(zeallot)
library(spdep)
prior_marker_info <- matrix(0,
                            nrow = length(CodexPredefined$classifier2$positive),
                            ncol = nrow(mergedSeu),
                            dimnames = list(
                              names(CodexPredefined$classifier2$positive),
                              rownames(mergedSeu)
                            ))
for(i in seq_along(CodexPredefined$classifier2$positive)){
  prior_marker_info[names(CodexPredefined$classifier2$positive)[i],
                    CodexPredefined$classifier2$positive[[i]]$symbol[
                      CodexPredefined$classifier2$positive[[i]]$symbol %in%
                        colnames(prior_marker_info)
                    ]] <- 1
}
## reorder the celltype by lineage level
prior_marker_info <- prior_marker_info[
  c('Tumour', 'Stromal', 'HSC',
    'NK', 'B cell', 'T cell', 'Monocyte',
    'CD4_TFH', 'CD8 T-cell', 'T-reg',
    'Macrophage', 'M2 macrophage', 'cd45+lcp1', 'Dendritic cell'), ]
prior_marker_info <- cbind(
  celltype = rownames(prior_marker_info),
  Lineage_level=paste(c(1, 1, 1,
                        2, 2, 2, 2,
                        3, 3, 3,
                        3, 3, 3, 3), ## lineage level
                      c(0, 0, 0,
                        3, 3, 3, 3,
                        6, 6, 6,
                        7, 7, 7, 7), ## lineage ancestor 
                      c(1, 2, 3,
                        4, 5, 6, 7,
                        8, 9, 10,
                        11, 12, 13, 14), ## index
                      sep='_'),
  data.frame(prior_marker_info))
mergedSeu <- JoinLayers(mergedSeu)
imageData <- GetAssayData(mergedSeu, assay = 'RNA', layer = 'data')
imageData <- cbind(mergedSeu[[]], t(imageData))
colnames(imageData)[colnames(imageData)=='Cell.X'] <- 'X'
colnames(imageData)[colnames(imageData)=='Cell.Y'] <- 'Y'
CelestaObj <- CreateCelestaObject(
  project_title = "test",
  prior_marker_info = prior_marker_info,
  imaging_data = imageData)
CelestaObj <- FilterCells(
  CelestaObj, 
  high_marker_threshold=0.9,
  low_marker_threshold=0.1)
CelestaObj <- AssignCells(
  CelestaObj,
  max_iteration=10,
  cell_change_threshold=0.01)
mergedSeu$celltypeCELESTA <- 
  CelestaObj@final_cell_type_assignment[, 'Final cell type']
dittoBarPlot(mergedSeu,
             var = 'celltypeCELESTA',
             group.by='orig.ident')
```


### Neighborhoods analysis

After annotating cell types, we can assess whether these annotations exhibit spatial enrichment and conduct analyses of cellular neighborhoods throughout the tissue. Here we borrow the power of `hoodscanR` to identifying interactions between spatial communities.

```{r neighborhoodsAna}
clusterK <- 12
spe <- neighborhoodsAna(mergedSeu, ## if no merging step, use `seu`.
                        clusterK = clusterK,
                        k = 10, ## maximal 10 neightbors
                        anno_col = 'celltype_first')
# Plot the metrics for probability matrix
plotTissue(spe, color = entropy) +
    scale_color_gradientn(colors=CodexPredefined$heatColor)
plotTissue(spe, color = perplexity) +
    scale_color_gradientn(colors=CodexPredefined$heatColor)
# Plot the clusters for the probability matrix
plotTissue(spe, color = clusters)
# Plot heatmap for neighbourhood analysis
plotColocal(spe, pm_cols = unique(spe$celltype_first))
# Plot probability distribution
plotProbDist(spe, pm_cols = unique(spe$celltype_first), 
             by_cluster = TRUE, plot_all = TRUE, 
             show_clusters = as.character(seq(clusterK)))
```

### Co-occurrence analysis

The Jaccard Index is a statistic used for gauging the similarity and diversity of sample sets. The Jaccard Index is defined as: $\frac{cooccurent\_celltype\_A\_and\_B}{count\_celltype\_A+count\_celltype\_B-cooccurent\_celltype\_A\_and\_B}$.

This method are based on simple yet intuitive measures, such as counting the number of nearest neighbors of every cell and inspecting the proportion of different cell categories in a given Euclidean (2 norm aka $L_2$) distance ($\sqrt{\sum_{i}{(x_i - y_i})^2}$, the shortest distance).

```{r}
## here we set two times of cell length as maximal gaps to check 
## the cell co-location
maxgap <- max(mergedSeu$Cell.Length.um) * 2
ji <- JaccardIndex(mergedSeu, maxRadius = maxgap, anno_col='celltype_first')
plotJaccardIndex(ji)
```


The co-occurrence score gives us an indication on whether given two factors such as cell cluster, cell type, and etc, co-occur with each other at specific distances across the tissue. The co-occurrence score is defined as: $\frac{p(factor1|factor2)}{p(factor1)*p(factor2)}$.

```{r coOccurence}
maxgap <- max(mergedSeu$Cell.Length.um) * 50
## precalculate cell distance to save time because we will use same maxgap for
## both coOccurrence and spatialScore
celldist <- findNearCellsByRadius(
      seu=mergedSeu,
      maxRadius=maxgap)
cooc <- coOccurrence(mergedSeu, maxRadius=maxgap, bin=50,
                     anno_col='celltype',
                     celldist=celldist)
ggplot(data.frame(x=seq.int(nrow(cooc))*20-10, y=cooc[, 'cell proportion']),
       aes(x, y)) + geom_line() + theme_classic() +
  labs(x='distance', y='proportion')
plotData <- reshape2::melt(cooc[, -1])
ggplot(plotData, aes(x=Var1, y=Var2, fill = value))+
  geom_tile() + labs(x='bin', y='') +
  scale_fill_gradient(low = "white", high = "red")
```
The p-value of co-occurence can be calculated based on permutation test or a more rigorous framework, the homogeneous multitype Poisson point process (PPP) or complete spatial randomness and independence (CSRI) by picking a specific value of the radius (r). To perform the analysis, you can refer to R-package [SpicyR](https://www.bioconductor.org/packages/spicyR/) or [SpaceANOVA](https://github.com/sealx017/SpaceANOVA).

The spatial score define the distance ratio among three factors. The spatial score is defined as: $\frac{celltypeCenter\_distance\_to\_nearest\_celltypeRight}{celltypeCenter\_distance\_to\_nearest\_celltypeLeft}$. The high value indicates the left celltype is closer than the right celltype.

```{r spatialScore}
## need to increase the neighbors number
ss <- spatialScore(mergedSeu,
                   celltypeCenter = "Macrophage",
                   celltypeLeft = "T cell",
                   celltypeRight = "Monocyte",
                   anno_col = 'celltype_first',
                   maxRadius=maxgap,
                   celldist=celldist)
head(ss)
boxplot(ss$spatialScore)
```

# Session Info
```{r sessionInfo, results='asis'}
sessionInfo()
```

