<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>cdQuPath Vignette: overview • cdQuPath</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="cdQuPath Vignette: overview">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">cdQuPath</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.0.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/cdQuPath.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jianhong/cdQuPath" class="external-link">
    <span class="fa fa-github"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>cdQuPath Vignette: overview</h1>
                        <h4 data-toc-skip class="author">Jianhong
Ou</h4>
            
            <h4 data-toc-skip class="date">20 August 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/jianhong/cdQuPath/blob/main/vignettes/cdQuPath.Rmd" class="external-link"><code>vignettes/cdQuPath.Rmd</code></a></small>
      <div class="hidden name"><code>cdQuPath.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This package presents a fundamental workflow for spatial phenotyping,
cell type annotation, multi-nucleus detection, quality control, and
calculating cell neighborhoods from whole slide images of ultra-high
multiplexed tissue captured by ‘Akoya Biosciences’ assays. The workflow
leverages QuPath, StarDist, Cellpose, Seurat, and hoodscanR.</p>
<p>The upstream code are available in the folder of
<code>extdata</code>. The detailed documentation will be provide in
following sections.</p>
<p>The <em>cdQuPath</em> aims to provide a easy to use tool for
downstream analysis with <code>Seurat</code> and <code>hoodscanR</code>
for quality control and neighborhoods analysis.</p>
</div>
<div class="section level2">
<h2 id="quick-start">Quick start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<p>Here is an example using <em>cdQuPath</em>.</p>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<p>First, please download and install <a href="https://qupath.github.io/" class="external-link">QuPath</a> (v0.4.0 or newer). The cell
segmentation requires the <a href="https://github.com/BIOP/qupath-extension-cellpose" class="external-link">cellpose</a>
and <a href="https://github.com/qupath/qupath-extension-stardist" class="external-link">stardist</a>
extension for <code>QuPath</code>. Don’t forget to set the Python path
for the <code>Cellpose</code> extension.</p>
<p>Then, download a copy of the Groovy script to your local machine.
There are two scripts available:
<code>cellpose_stardist_multiNucleis_underGUI.groovy</code> and
<code>ExportCellDetectionMeasurement.groovy</code>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">git</span> clone https://github.com/jianhong/cdQuPath.git</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="bu">cd</span> cdQuPath/inst/extdata/scripts</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fu">ls</span></span></code></pre></div>
<p>Then, install <em>cdQuPath</em> and other packages required to run
the examples.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">BiocManager</a></span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"jianhong/cdQuPath"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-1--cell-segmentation">Step 1. Cell segmentation<a class="anchor" aria-label="anchor" href="#step-1--cell-segmentation"></a>
</h3>
<p>This step involves merging the results from <a href="https://www.cellpose.org/" class="external-link">Cellpose</a> and <a href="https://github.com/stardist/stardist" class="external-link">Stardist</a> analyses.</p>
<p>Navigate to line 25 of the
<code>cellpose_stardist_multiNucleis_underGUI.groovy</code> script and
update the path of the <code>stardistModel</code> variable to match the
file path of your <code>cell_seg_model.pb</code>. Adjust the parameters
from <a href="https://github.com/jianhong/codex_qupath/blob/main/scr/cellpose_stardist_multiNucleis_underGUI.groovy#L26-L38" class="external-link">line
26 to line 38</a> accordingly. You may want to change the
<code>cytoChannels</code> and <code>measurementPercentileCutoff</code>
according to your observations. The <code>distanceCutoff</code> is the
maximal cutoff distance of nucleus for multiple nucleus cells.</p>
<p>Then, select a region in the opened TIFF view window in QuPath. Start
with a small region when testing the script, and expand it once you
achieve the desired results. Please note that if you observe regional
biases, you may want to exclude the biases regions. If it is
un-avoidable, you may want to normalize the signals. We will discuss
this later.</p>
<p>Finally, run the
<code>cellpose_stardist_multiNucleis_underGUI.groovy</code> script by
clicking the Run button in the Script Editor.</p>
<p><a href="https://youtu.be/3SKKSDhlEkw" class="external-link"><img src="https://img.youtube.com/vi/3SKKSDhlEkw/0.jpg" alt="Youtube Video"></a></p>
</div>
<div class="section level3">
<h3 id="step-2--export-the-cell-segmentation">Step 2. Export the cell segmentation<a class="anchor" aria-label="anchor" href="#step-2--export-the-cell-segmentation"></a>
</h3>
<p>The script <code>ExportCellDetectionMeasurement.groovy</code> is
designed to export cell segmentation data. Upon export, two files are
generated and saved in a folder prefixed with
<code>measurementsExport</code>.</p>
<ul>
<li>The <code>.tsv</code> file is compatible with
<code>createSeuratObj.R</code>.</li>
<li>The <code>.geojson</code> file is intended for reloading into
<code>QuPath</code>.</li>
</ul>
<p>These exported files contain comprehensive information, including
cell area, locations, marker signal statistics, and nucleus
classification. The cell locations are particularly useful for
neighborhood analysis.</p>
</div>
<div class="section level3">
<h3 id="step-3--work-in-r">Step 3. Work in R<a class="anchor" aria-label="anchor" href="#step-3--work-in-r"></a>
</h3>
<div class="section level4">
<h4 id="load-library">Load library<a class="anchor" aria-label="anchor" href="#load-library"></a>
</h4>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jianhong/cdQuPath" class="external-link">cdQuPath</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">dittoSeq</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/DavisLaboratory/hoodscanR" class="external-link">hoodscanR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drighelli/SpatialExperiment" class="external-link">SpatialExperiment</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="create-a-seurat-object-from-output-tsv-file-">Create a Seurat object from output <code>tsv</code> file.<a class="anchor" aria-label="anchor" href="#create-a-seurat-object-from-output-tsv-file-"></a>
</h4>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tsv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"sample.qptiff.tsv.gz"</span>,</span>
<span>                   package <span class="op">=</span> <span class="st">"cdQuPath"</span>,</span>
<span>                   mustWork <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createSeuratObj.html">createSeuratObj</a></span><span class="op">(</span><span class="va">tsv</span>, useValue <span class="op">=</span> <span class="st">'Median'</span>,</span>
<span>                       markerLocations <span class="op">=</span> <span class="va">CodexPredefined</span><span class="op">$</span><span class="va">markerLocations</span><span class="op">)</span></span>
<span><span class="va">seu</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 600 features across 1093 samples within 25 assays </span></span>
<span><span class="co">## Active assay: RNA (24 features, 0 variable features)</span></span>
<span><span class="co">##  1 layer present: counts</span></span>
<span><span class="co">##  24 other assays present: Nucleus_Mean, Nucleus_Median, Nucleus_Min, Nucleus_Max, Nucleus_Std.Dev., Nucleus_Variance, Cytoplasm_Mean, Cytoplasm_Median, Cytoplasm_Min, Cytoplasm_Max, Cytoplasm_Std.Dev., Cytoplasm_Variance, Membrane_Mean, Membrane_Median, Membrane_Min, Membrane_Max, Membrane_Std.Dev., Membrane_Variance, Cell_Mean, Cell_Median, Cell_Min, Cell_Max, Cell_Std.Dev., Cell_Variance</span></span>
<span><span class="co">##  1 dimensional reduction calculated: pos</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="quality-control">Quality Control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h4>
<p>The conventional method for evaluating dynamic range involves
computing a signal-to-background (SNR) ratio. This is done by dividing
the average intensity of the top 20 brightest cells by the average
intensity of the weakest 10% of cells. An SNR value of 10 or higher is
indicative of reliable image analysis. It is recommended that the SNR
falls within the range of (10, Inf], typically exceeding 100. An SNR
below 3 suggests that the antibodies are performing poorly.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## please make sure the default assay is 'RNA'</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "RNA"</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dynamicRanges.html">dynamicRanges</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span></span>
<span><span class="va">dr</span></span></code></pre></div>
<pre><code><span><span class="co">## $brightest</span></span>
<span><span class="co">##  DAPI.01    CD163    CD127     sox9     CD14      CD4      mpo     cd68 </span></span>
<span><span class="co">##  109.625   13.300   11.600   24.300   26.875   52.150   10.000  238.750 </span></span>
<span><span class="co">##     cd45    foxp3     lcp1      pd1      cd8    hif1a     ki67     cd19 </span></span>
<span><span class="co">##   19.700    8.750   18.350    7.400   10.525   30.575   80.625   17.650 </span></span>
<span><span class="co">##    cd206    znf90     asma      cd3     cd86 hist3h2a     cd34    CD11c </span></span>
<span><span class="co">##   12.550   34.850   20.875    2.150   11.925  105.600   50.000   42.000 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $weakest</span></span>
<span><span class="co">##   DAPI.01     CD163     CD127      sox9      CD14       CD4       mpo      cd68 </span></span>
<span><span class="co">## 47.686992  2.148882  2.790650  0.000000  2.899390  5.625508  3.911585 69.067073 </span></span>
<span><span class="co">##      cd45     foxp3      lcp1       pd1       cd8     hif1a      ki67      cd19 </span></span>
<span><span class="co">##  4.411077  3.632622  6.510163  2.408537  3.322663  2.966463  3.584858 10.713415 </span></span>
<span><span class="co">##     cd206     znf90      asma       cd3      cd86  hist3h2a      cd34     CD11c </span></span>
<span><span class="co">##  6.490854 17.210874  0.000000  0.000000  9.333841 23.114329 10.100610 23.494411 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $dynamic_range</span></span>
<span><span class="co">##   DAPI.01     CD163     CD127      sox9      CD14       CD4       mpo      cd68 </span></span>
<span><span class="co">##  2.298845  6.189265  4.156737       Inf  9.269190  9.270274  2.556508  3.456785 </span></span>
<span><span class="co">##      cd45     foxp3      lcp1       pd1       cd8     hif1a      ki67      cd19 </span></span>
<span><span class="co">##  4.466029  2.408728  2.818670  3.072405  3.167640 10.306886 22.490432  1.647467 </span></span>
<span><span class="co">##     cd206     znf90      asma       cd3      cd86  hist3h2a      cd34     CD11c </span></span>
<span><span class="co">##  1.933490  2.024883       Inf       Inf  1.277609  4.568595  4.950196  1.787659</span></span></code></pre>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## markers may not work;</span></span>
<span><span class="co">## Please note that DAPI.01 is used for cell detection.</span></span>
<span><span class="va">avoid_markers</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">dr</span><span class="op">$</span><span class="va">dynamic_range</span><span class="op">)</span><span class="op">[</span><span class="va">dr</span><span class="op">$</span><span class="va">dynamic_range</span><span class="op">&lt;</span><span class="fl">3</span> <span class="op">|</span> <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">dr</span><span class="op">$</span><span class="va">dynamic_range</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">avoid_markers</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "DAPI.01" "mpo"     "foxp3"   "lcp1"    "cd19"    "cd206"   "znf90"  </span></span>
<span><span class="co">## [8] "cd86"    "CD11c"</span></span></code></pre>
<p>Verify the cell area and nucleus area for multi-nuclei.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualize QC metrics as a violin plot</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">seu</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "orig.ident"              "nCount_RNA"             </span></span>
<span><span class="co">##  [3] "nFeature_RNA"            "Cell.classification"    </span></span>
<span><span class="co">##  [5] "Cell.X"                  "Cell.Y"                 </span></span>
<span><span class="co">##  [7] "Nucleus.Area.um.2"       "Nucleus.Length.um"      </span></span>
<span><span class="co">##  [9] "Nucleus.Circularity"     "Nucleus.Solidity"       </span></span>
<span><span class="co">## [11] "Nucleus.Max.diameter.um" "Nucleus.Min.diameter.um"</span></span>
<span><span class="co">## [13] "Cell.Area.um.2"          "Cell.Length.um"         </span></span>
<span><span class="co">## [15] "Cell.Circularity"        "Cell.Solidity"          </span></span>
<span><span class="co">## [17] "Cell.Max.diameter.um"    "Cell.Min.diameter.um"   </span></span>
<span><span class="co">## [19] "Nucleus.Cell.area.ratio" "Cell.isMultiNucleis"</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">VlnPlot</a></span><span class="op">(</span><span class="va">seu</span>,</span>
<span>        features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Cell.Area.um.2"</span>,</span>
<span>                     <span class="st">"Nucleus.Area.um.2"</span>,</span>
<span>                     <span class="st">'Nucleus.Cell.area.ratio'</span><span class="op">)</span>,</span>
<span>        split.by <span class="op">=</span> <span class="st">'Cell.isMultiNucleis'</span>,</span>
<span>        pt.size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">500</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>        ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="cdQuPath_files/figure-html/vlnplot1-1.png" width="700"></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## if you want to save the plot, try</span></span>
<span><span class="co">## ggsave(filename)</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">VlnPlot</a></span><span class="op">(</span><span class="va">seu</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'cd68'</span>, <span class="st">'CD11c'</span>, <span class="st">'cd34'</span><span class="op">)</span>,</span>
<span>        pt.size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">500</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>        ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="cdQuPath_files/figure-html/vlnPlot2-1.png" width="700"></p>
<p>Now, we can proceed with filtering the cells. Upon observation, we
notice that the average area of single-nucleus cells is below 500.
Therefore, we establish the cutoff for cell area at approximately three
times the size of single-nucleus cells. Additionally, if the
nucleus-to-cell area ratio is excessively high, it suggests potential
issues with the cell area detection process. We will filter out those
cells.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="va">seu</span><span class="op">$</span><span class="va">Cell.Area.um.2</span><span class="op">&lt;</span><span class="fl">1500</span> <span class="op">&amp;</span> <span class="va">seu</span><span class="op">$</span><span class="va">Nucleus.Cell.area.ratio</span> <span class="op">&lt;</span> <span class="fl">.9</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">seu</span>,</span>
<span>              cells <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span><span class="op">[</span><span class="va">keep</span><span class="op">]</span>,</span>
<span>              features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">dr</span><span class="op">$</span><span class="va">dynamic_range</span><span class="op">)</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">dr</span><span class="op">$</span><span class="va">dynamic_range</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">seu</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 600 features across 1088 samples within 25 assays </span></span>
<span><span class="co">## Active assay: RNA (24 features, 0 variable features)</span></span>
<span><span class="co">##  1 layer present: counts</span></span>
<span><span class="co">##  24 other assays present: Nucleus_Mean, Nucleus_Median, Nucleus_Min, Nucleus_Max, Nucleus_Std.Dev., Nucleus_Variance, Cytoplasm_Mean, Cytoplasm_Median, Cytoplasm_Min, Cytoplasm_Max, Cytoplasm_Std.Dev., Cytoplasm_Variance, Membrane_Mean, Membrane_Median, Membrane_Min, Membrane_Max, Membrane_Std.Dev., Membrane_Variance, Cell_Mean, Cell_Median, Cell_Min, Cell_Max, Cell_Std.Dev., Cell_Variance</span></span>
<span><span class="co">##  1 dimensional reduction calculated: pos</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="normalizing-the-data">Normalizing the data<a class="anchor" aria-label="anchor" href="#normalizing-the-data"></a>
</h4>
<p>The initial query is whether we are observing abnormally high or low
intensity regions in specific sub-regions. If these variations are
deemed unrelated to tissue-specific characteristics, it might be
advisable to normalize each marker channel accordingly. The positional
data of cells is contained within the reduction matrix labeled
‘pos’.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 24</span></span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">seu</span>, slot<span class="op">=</span><span class="st">'count'</span>,</span>
<span>            features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>            reduction <span class="op">=</span> <span class="st">'pos'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_reverse</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="cdQuPath_files/figure-html/regionCheck-1.png" width="700"></p>
<p>You may attempt to normalize the irregular sub-regions using LOESS
(Local Polynomial Regression Fitting) with the <code>loessSmooth</code>
function.</p>
<p>The next step involves cell-wise normalization. The available
normalization methods are ‘LogNormalize’, ‘CLR’ (centered log ratio),
‘RC’ (CPM), and ‘zscore’, which can be applied using the
<code>normData</code> function. The normalization method except for
‘zscore’ is borrowed from <code>Seurat</code> package.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normData.html">normData</a></span><span class="op">(</span><span class="va">seu</span>, normalizationMethod<span class="op">=</span><span class="st">'zscore'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="identification-of-highly-variable-markers-marker-selection">Identification of highly variable markers (marker selection)<a class="anchor" aria-label="anchor" href="#identification-of-highly-variable-markers-marker-selection"></a>
</h4>
<p>In the background, this step will execute the
<code>FindVariableFeatures</code> function and exclude the
‘avoidMarkers’ from the list of variable markers. Please ensure that
‘nFeatures’ is smaller than the total number of available markers.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findVarMarkers.html">findVarMarkers</a></span><span class="op">(</span></span>
<span>    <span class="va">seu</span>,</span>
<span>    avoidMarkers<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'DAPI.01'</span>, <span class="st">'hist3h2a'</span><span class="op">)</span>,</span>
<span>    nFeatures <span class="op">=</span> <span class="fl">10L</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "ki67"  "CD4"   "cd34"  "hif1a" "mpo"   "znf90" "cd19"  "CD127" "CD14"</span></span></code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/VariableFeaturePlot.html" class="external-link">VariableFeaturePlot</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/LabelPoints.html" class="external-link">LabelPoints</a></span><span class="op">(</span>plot <span class="op">=</span> <span class="va">plot1</span>, points <span class="op">=</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span>, repel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="cdQuPath_files/figure-html/variableFeatures-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="scaling-the-data-and-do-cluster-analysis">Scaling the data and do cluster analysis<a class="anchor" aria-label="anchor" href="#scaling-the-data-and-do-cluster-analysis"></a>
</h4>
<p>The following steps outline the standard process of single-cell
analysis in the <code>Seurat</code> package. You may adjust the process
accordingly to suit your needs.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">all.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">seu</span>, features <span class="op">=</span> <span class="va">all.markers</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">seu</span>, features <span class="op">=</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">dim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq.int</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Embeddings.html" class="external-link">Embeddings</a></span><span class="op">(</span><span class="va">seu</span><span class="op">$</span><span class="va">pca</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/VizDimLoadings.html" class="external-link">VizDimLoadings</a></span><span class="op">(</span><span class="va">seu</span>, dims <span class="op">=</span> <span class="va">dim</span>,</span>
<span>               nfeatures<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span><span class="op">)</span>,</span>
<span>               reduction <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span></span></code></pre></div>
<p><img src="cdQuPath_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">seu</span>, dims <span class="op">=</span> <span class="va">dim</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span><span class="va">seu</span>, resolution <span class="op">=</span> <span class="fl">0.9</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of nodes: 1088</span></span>
<span><span class="co">## Number of edges: 32836</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Running Louvain algorithm...</span></span>
<span><span class="co">## Maximum modularity in 10 random starts: 0.7546</span></span>
<span><span class="co">## Number of communities: 10</span></span>
<span><span class="co">## Elapsed time: 0 seconds</span></span></code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">seu</span>, reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="va">dim</span><span class="op">)</span></span>
<span><span class="va">pt.size</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">500</span>, <span class="fl">0</span>, <span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seu</span>,</span>
<span>        reduction <span class="op">=</span> <span class="st">'umap'</span>, label <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>        pt.size <span class="op">=</span> <span class="va">pt.size</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_reverse</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="cdQuPath_files/figure-html/unnamed-chunk-3-2.png" width="700"></p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## check the distribution of the clusters in spatial position.</span></span>
<span><span class="va">ncol</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">ceiling</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">seu</span><span class="op">$</span><span class="va">seurat_clusters</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seu</span>,</span>
<span>        reduction <span class="op">=</span> <span class="st">'pos'</span>, label <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>        split.by <span class="op">=</span> <span class="st">'seurat_clusters'</span>,</span>
<span>        ncol <span class="op">=</span> <span class="va">ncol</span>, pt.size <span class="op">=</span> <span class="va">pt.size</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_reverse</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="cdQuPath_files/figure-html/unnamed-chunk-3-3.png" width="700"></p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## check the nucleis distribution</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/dittoSeq/man/dittoBarPlot.html" class="external-link">dittoBarPlot</a></span><span class="op">(</span><span class="va">seu</span>,</span>
<span>             var <span class="op">=</span> <span class="st">'Cell.isMultiNucleis'</span>,</span>
<span>             group.by<span class="op">=</span><span class="st">'seurat_clusters'</span><span class="op">)</span></span></code></pre></div>
<p><img src="cdQuPath_files/figure-html/unnamed-chunk-3-4.png" width="700"></p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seu</span>, reduction <span class="op">=</span> <span class="st">'pos'</span>, label <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>        split.by <span class="op">=</span> <span class="st">'Cell.isMultiNucleis'</span>,</span>
<span>        pt.size <span class="op">=</span> <span class="va">pt.size</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_reverse</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="cdQuPath_files/figure-html/unnamed-chunk-3-5.png" width="700"></p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">VlnPlot</a></span><span class="op">(</span><span class="va">seu</span>,</span>
<span>        features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Nucleus.Area.um.2'</span>,</span>
<span>                     <span class="st">'Cell.Area.um.2'</span>,</span>
<span>                     <span class="st">'Nucleus.Cell.area.ratio'</span><span class="op">)</span>,</span>
<span>        pt.size <span class="op">=</span> <span class="va">pt.size</span><span class="op">)</span></span></code></pre></div>
<p><img src="cdQuPath_files/figure-html/unnamed-chunk-3-6.png" width="700"></p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Dot plot</span></span>
<span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">FindAllMarkers</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DotPlot.html" class="external-link">DotPlot</a></span><span class="op">(</span><span class="va">seu</span>, features <span class="op">=</span> <span class="va">markers</span>, scale <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="cdQuPath_files/figure-html/unnamed-chunk-3-7.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="rescale-data-using-gaussian-mixture-model">Rescale data using Gaussian Mixture Model<a class="anchor" aria-label="anchor" href="#rescale-data-using-gaussian-mixture-model"></a>
</h4>
<p>In the downstream analysis, cell types are detected by fitting a
multiple-class Gaussian Mixture Model (GMM) to the re-scaled data. The
raw intensities are converted to probabilities, and a cutoff of 0.5 is
set for cell type assignment.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitGMM.html">fitGMM</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DotPlot.html" class="external-link">DotPlot</a></span><span class="op">(</span><span class="va">seu</span>, features <span class="op">=</span> <span class="va">markers</span>, assay <span class="op">=</span> <span class="st">'GMM'</span>, scale <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="cdQuPath_files/figure-html/gmm-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="correlation-heatmap">Correlation heatmap<a class="anchor" aria-label="anchor" href="#correlation-heatmap"></a>
</h4>
<p>The correlation heatmap will display the signal similarity among
different markers. The distance will be calculated using the
<code><a href="https://rdrr.io/r/stats/dist.html" class="external-link">stats::dist</a></code> function. Users can experiment with different
inputs such as z-scales, GMM data, and so on.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "RNA"</span></span></code></pre>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/markerCorrelation.html">markerCorrelation</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "counts"     "data"       "scale.data"</span></span></code></pre>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## plot the correlations for data, here it is Z-scale data</span></span>
<span><span class="va">p</span><span class="op">[[</span><span class="st">'data'</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="cdQuPath_files/figure-html/cor-1.png" width="700"></p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## use GMM rescaled data to plot the correlation</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">'GMM'</span></span>
<span><span class="va">p_GMM</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/markerCorrelation.html">markerCorrelation</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span></span>
<span><span class="va">p_GMM</span><span class="op">[[</span><span class="st">'data'</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="cdQuPath_files/figure-html/cor-2.png" width="700"></p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## </span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">'RNA'</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="assign-cell-types">Assign cell types<a class="anchor" aria-label="anchor" href="#assign-cell-types"></a>
</h4>
<p>The phenotype will utilize the re-scaled data obtained through GMM.
Two methods are available for assigning cell types: ‘Rank’ and
‘Boolean’. With the ‘Rank’ option, cell types are assigned based on the
rank of probabilities for the positive markers. On the other hand, the
‘Boolean’ option assigns cell types according to the order of the
classifier for the positive markers.</p>
<p>The metadata ‘celltype’ stores the assigned cell types for each
Seurat cluster. Additionally, the metadata ‘celltype_all’ contains all
possible cell types for each cell, determined by a cutoff of 0.5.
Finally, the metadata ‘celltype_first’ represents the initial
assignment.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/detectCellType.html">detectCellType</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">seu</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##              B CELLS    Cytotoxic T Cells          Endothelial </span></span>
<span><span class="co">##                  220                  163                   87 </span></span>
<span><span class="co">##       Helper T cells        M1 Macrophage        M2 MACROPHAGE </span></span>
<span><span class="co">##                  162                  114                  147 </span></span>
<span><span class="co">##             MONOCYTE Proliferation Marker </span></span>
<span><span class="co">##                   47                  148</span></span></code></pre>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## check the cell type annotation by clusters</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seu</span>, group.by <span class="op">=</span> <span class="st">'celltype'</span><span class="op">)</span></span></code></pre></div>
<p><img src="cdQuPath_files/figure-html/celltype-1.png" width="700"></p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seu</span>, group.by <span class="op">=</span> <span class="st">'celltype'</span>, reduction <span class="op">=</span> <span class="st">'pos'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_reverse</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="cdQuPath_files/figure-html/celltype-2.png" width="700"></p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/dittoSeq/man/dittoBarPlot.html" class="external-link">dittoBarPlot</a></span><span class="op">(</span><span class="va">seu</span>,</span>
<span>             var <span class="op">=</span> <span class="st">'celltype'</span>,</span>
<span>             group.by<span class="op">=</span><span class="st">'Cell.isMultiNucleis'</span><span class="op">)</span></span></code></pre></div>
<p><img src="cdQuPath_files/figure-html/celltype-3.png" width="700"></p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## check the first annotation of each cell, not restricted by seurat_cluster</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seu</span>, group.by <span class="op">=</span> <span class="st">'celltype_first'</span>, reduction <span class="op">=</span> <span class="st">'pos'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_reverse</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="cdQuPath_files/figure-html/celltype-4.png" width="700"></p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/dittoSeq/man/dittoBarPlot.html" class="external-link">dittoBarPlot</a></span><span class="op">(</span><span class="va">seu</span>,</span>
<span>             var <span class="op">=</span> <span class="st">'celltype_first'</span>,</span>
<span>             group.by<span class="op">=</span><span class="st">'Cell.isMultiNucleis'</span><span class="op">)</span></span></code></pre></div>
<p><img src="cdQuPath_files/figure-html/celltype-5.png" width="700"></p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## check the signals for each celltype</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">'celltype'</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DotPlot.html" class="external-link">DotPlot</a></span><span class="op">(</span><span class="va">seu</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span>, scale <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="cdQuPath_files/figure-html/celltype-6.png" width="700"></p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## export the celltype</span></span>
<span><span class="va">ct</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>Object.ID<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span>,</span>
<span>                 x<span class="op">=</span><span class="va">seu</span><span class="op">$</span><span class="va">Cell.X</span>,</span>
<span>                 y<span class="op">=</span><span class="va">seu</span><span class="op">$</span><span class="va">Cell.Y</span>,</span>
<span>                 CellType<span class="op">=</span><span class="va">seu</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ct</span>, n<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                                                 Object.ID</span></span>
<span><span class="co">## 8d5ccdac-31da-4930-81b6-dd9f272c10d7 8d5ccdac-31da-4930-81b6-dd9f272c10d7</span></span>
<span><span class="co">## b21deea4-2d9c-4e0b-9c83-591622b33a9a b21deea4-2d9c-4e0b-9c83-591622b33a9a</span></span>
<span><span class="co">##                                             x        y       CellType</span></span>
<span><span class="co">## 8d5ccdac-31da-4930-81b6-dd9f272c10d7 13142.86 8012.715 Helper T cells</span></span>
<span><span class="co">## b21deea4-2d9c-4e0b-9c83-591622b33a9a 13142.86 8012.715 Helper T cells</span></span></code></pre>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## if you see postfix like _1, then run</span></span>
<span><span class="va">ct</span><span class="op">$</span><span class="va">Object.ID</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">'_\\d+$'</span>, <span class="st">''</span>, <span class="va">ct</span><span class="op">$</span><span class="va">Object.ID</span><span class="op">)</span></span>
<span><span class="co">## output to same folder: celltype_csv &lt;- sub('qptiff.tsv', 'celltype.csv', tsv)</span></span>
<span><span class="va">celltype_csv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>fileext <span class="op">=</span> <span class="st">'.csv'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.csv</a></span><span class="op">(</span><span class="va">ct</span>, file <span class="op">=</span> <span class="va">celltype_csv</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">## after exporting, you can load the celltypes back to QuPath by running the script load_celltype.groovy in the scripts folder.</span></span></code></pre></div>
<p><a href="https://youtu.be/OH8TKCxZKkM" class="external-link"><img src="https://img.youtube.com/vi/OH8TKCxZKkM/0.jpg" alt="Youtube Video"></a></p>
</div>
<div class="section level4">
<h4 id="merge-multiple-seurat-objects-if-necessary">Merge multiple Seurat objects if necessary<a class="anchor" aria-label="anchor" href="#merge-multiple-seurat-objects-if-necessary"></a>
</h4>
<p>If you have datasets from multiple sub-regions, you can process each
one individually as shown above and then merge them. Occasionally, some
images may not have the same background across different regions. To
address this, you can detect the cells in each sub-region with a similar
background or baseline and then perform GMM rescaling for each
sub-region. Once this is done, you can combine the analyzed results to
create a merged dataset.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tsv2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"sample2.qptiff.tsv.gz"</span>,</span>
<span>                   package <span class="op">=</span> <span class="st">"cdQuPath"</span>,</span>
<span>                   mustWork <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">seu2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createSeuratObj.html">createSeuratObj</a></span><span class="op">(</span><span class="va">tsv2</span>, useValue <span class="op">=</span> <span class="st">'Median'</span>,</span>
<span>                       markerLocations <span class="op">=</span> <span class="va">CodexPredefined</span><span class="op">$</span><span class="va">markerLocations</span><span class="op">)</span></span>
<span><span class="va">seu2</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 600 features across 282 samples within 25 assays </span></span>
<span><span class="co">## Active assay: RNA (24 features, 0 variable features)</span></span>
<span><span class="co">##  1 layer present: counts</span></span>
<span><span class="co">##  24 other assays present: Nucleus_Mean, Nucleus_Median, Nucleus_Min, Nucleus_Max, Nucleus_Std.Dev., Nucleus_Variance, Cytoplasm_Mean, Cytoplasm_Median, Cytoplasm_Min, Cytoplasm_Max, Cytoplasm_Std.Dev., Cytoplasm_Variance, Membrane_Mean, Membrane_Median, Membrane_Min, Membrane_Max, Membrane_Std.Dev., Membrane_Variance, Cell_Mean, Cell_Median, Cell_Min, Cell_Max, Cell_Std.Dev., Cell_Variance</span></span>
<span><span class="co">##  1 dimensional reduction calculated: pos</span></span></code></pre>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="va">seu2</span><span class="op">$</span><span class="va">Cell.Area.um.2</span><span class="op">&lt;</span><span class="fl">1500</span> <span class="op">&amp;</span> <span class="va">seu2</span><span class="op">$</span><span class="va">Nucleus.Cell.area.ratio</span> <span class="op">&lt;</span> <span class="fl">.9</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">keep</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## keep</span></span>
<span><span class="co">## FALSE  TRUE </span></span>
<span><span class="co">##     1   281</span></span></code></pre>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">seu2</span>,</span>
<span>              cells <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">seu2</span><span class="op">)</span><span class="op">[</span><span class="va">keep</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">seu2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normData.html">normData</a></span><span class="op">(</span><span class="va">seu2</span>, normalizationMethod<span class="op">=</span><span class="st">'zscore'</span><span class="op">)</span></span>
<span><span class="va">seu2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findVarMarkers.html">findVarMarkers</a></span><span class="op">(</span></span>
<span>    <span class="va">seu2</span>,</span>
<span>    avoidMarkers<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'DAPI.01'</span>, <span class="st">'hist3h2a'</span><span class="op">)</span>,</span>
<span>    nFeatures <span class="op">=</span> <span class="fl">10L</span><span class="op">)</span></span>
<span><span class="va">seu2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">seu2</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">seu2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">seu2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">seu2</span>, features <span class="op">=</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span> <span class="va">seu2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">dim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq.int</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Embeddings.html" class="external-link">Embeddings</a></span><span class="op">(</span><span class="va">seu2</span><span class="op">$</span><span class="va">pca</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">seu2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">seu2</span>, dims <span class="op">=</span> <span class="va">dim</span><span class="op">)</span></span>
<span><span class="va">seu2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span><span class="va">seu2</span>, resolution <span class="op">=</span> <span class="fl">0.9</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of nodes: 281</span></span>
<span><span class="co">## Number of edges: 8053</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Running Louvain algorithm...</span></span>
<span><span class="co">## Maximum modularity in 10 random starts: 0.6137</span></span>
<span><span class="co">## Number of communities: 5</span></span>
<span><span class="co">## Elapsed time: 0 seconds</span></span></code></pre>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitGMM.html">fitGMM</a></span><span class="op">(</span><span class="va">seu2</span><span class="op">)</span></span>
<span><span class="va">seu2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/detectCellType.html">detectCellType</a></span><span class="op">(</span><span class="va">seu2</span><span class="op">)</span></span>
<span><span class="va">seu2</span><span class="op">$</span><span class="va">pca</span> <span class="op">&lt;-</span> <span class="cn">NULL</span> <span class="co"># to advoid different pca columns error.</span></span>
<span><span class="va">seu</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">&lt;-</span> <span class="st">'region1'</span></span>
<span><span class="va">seu2</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">&lt;-</span> <span class="st">'region2'</span></span>
<span><span class="va">mergedSeu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">seu</span>, y<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">seu2</span><span class="op">)</span>, merge.dr <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">mergedSeu</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 624 features across 1369 samples within 26 assays </span></span>
<span><span class="co">## Active assay: RNA (24 features, 8 variable features)</span></span>
<span><span class="co">##  6 layers present: counts.1, counts.2, data.1, scale.data.1, data.2, scale.data.2</span></span>
<span><span class="co">##  25 other assays present: Nucleus_Mean, Nucleus_Median, Nucleus_Min, Nucleus_Max, Nucleus_Std.Dev., Nucleus_Variance, Cytoplasm_Mean, Cytoplasm_Median, Cytoplasm_Min, Cytoplasm_Max, Cytoplasm_Std.Dev., Cytoplasm_Variance, Membrane_Mean, Membrane_Median, Membrane_Min, Membrane_Max, Membrane_Std.Dev., Membrane_Variance, Cell_Mean, Cell_Median, Cell_Min, Cell_Max, Cell_Std.Dev., Cell_Variance, GMM</span></span>
<span><span class="co">##  3 dimensional reductions calculated: pos, pca, umap</span></span></code></pre>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/ObjectAccess.html" class="external-link">Assays</a></span><span class="op">(</span><span class="va">mergedSeu</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class "SimpleAssays"</span></span>
<span><span class="co">## Slot "data":</span></span>
<span><span class="co">## List of length 1</span></span></code></pre>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/dittoSeq/man/dittoBarPlot.html" class="external-link">dittoBarPlot</a></span><span class="op">(</span><span class="va">mergedSeu</span>,</span>
<span>             var <span class="op">=</span> <span class="st">'celltype_first'</span>,</span>
<span>             group.by<span class="op">=</span><span class="st">'orig.ident'</span><span class="op">)</span></span></code></pre></div>
<p><img src="cdQuPath_files/figure-html/runSample2-1.png" width="700"></p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">mergedSeu</span>, slot<span class="op">=</span><span class="st">'count'</span>,</span>
<span>            features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">mergedSeu</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>            reduction <span class="op">=</span> <span class="st">'pos'</span>,</span>
<span>            ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_reverse</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="cdQuPath_files/figure-html/runSample2-2.png" width="700"></p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">mergedSeu</span>, group.by <span class="op">=</span> <span class="st">'celltype'</span>, reduction <span class="op">=</span> <span class="st">'pos'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_reverse</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="cdQuPath_files/figure-html/runSample2-3.png" width="700"></p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## use GMM rescaled data to plot the correlation</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">mergedSeu</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">'GMM'</span></span>
<span><span class="va">p_GMM</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/markerCorrelation.html">markerCorrelation</a></span><span class="op">(</span><span class="va">mergedSeu</span>, layers <span class="op">=</span> <span class="st">'data'</span><span class="op">)</span></span>
<span><span class="va">p_GMM</span><span class="op">[[</span><span class="st">'data'</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="cdQuPath_files/figure-html/runSample2-4.png" width="700"></p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## </span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">mergedSeu</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">'RNA'</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="detect-the-celltype-by-clissifier">Detect the celltype by clissifier<a class="anchor" aria-label="anchor" href="#detect-the-celltype-by-clissifier"></a>
</h4>
<p>You can also try to detect the cell type by a generalized k-Nearest
Neighbors Classifier.</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">classifier</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'classifier.CSA.res0.3.rds'</span>, </span>
<span>                      package<span class="op">=</span><span class="st">'cdQuPath'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## this request lots of memory.</span></span>
<span><span class="va">mergedSeu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/detectCellType.html">detectCellType</a></span><span class="op">(</span><span class="va">mergedSeu</span>,</span>
<span>               classifier<span class="op">=</span><span class="va">classifier</span>,</span>
<span>               celltypeColumnName<span class="op">=</span><span class="st">'celltypeByKNN'</span><span class="op">)</span></span>
<span><span class="co">## this does not work for test dataset.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">mergedSeu</span><span class="op">$</span><span class="va">celltypeByKNN</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="neighborhoods-analysis">Neighborhoods analysis<a class="anchor" aria-label="anchor" href="#neighborhoods-analysis"></a>
</h4>
<p>After annotating cell types, we can assess whether these annotations
exhibit spatial enrichment and conduct analyses of cellular
neighborhoods throughout the tissue. Here we borrow the power of
<code>hoodscanR</code> to identifying interactions between spatial
communities.</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clusterK</span> <span class="op">&lt;-</span> <span class="fl">12</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/neighborhoodsAna.html">neighborhoodsAna</a></span><span class="op">(</span><span class="va">mergedSeu</span>, <span class="co">## if no merging step, use `seu`.</span></span>
<span>                        clusterK <span class="op">=</span> <span class="va">clusterK</span>,</span>
<span>                        k <span class="op">=</span> <span class="fl">10</span>, <span class="co">## maximal 10 neightbors</span></span>
<span>                        anno_col <span class="op">=</span> <span class="st">'celltype_first'</span><span class="op">)</span></span>
<span><span class="co"># Plot the metrics for probability matrix</span></span>
<span><span class="fu"><a href="https://davislaboratory.github.io/hoodscanR/reference/plotTissue.html" class="external-link">plotTissue</a></span><span class="op">(</span><span class="va">spe</span>, color <span class="op">=</span> <span class="va">entropy</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradientn</a></span><span class="op">(</span>colors<span class="op">=</span><span class="va">CodexPredefined</span><span class="op">$</span><span class="va">heatColor</span><span class="op">)</span></span></code></pre></div>
<p><img src="cdQuPath_files/figure-html/neighborhoodsAna-1.png" width="700"></p>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://davislaboratory.github.io/hoodscanR/reference/plotTissue.html" class="external-link">plotTissue</a></span><span class="op">(</span><span class="va">spe</span>, color <span class="op">=</span> <span class="va">perplexity</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradientn</a></span><span class="op">(</span>colors<span class="op">=</span><span class="va">CodexPredefined</span><span class="op">$</span><span class="va">heatColor</span><span class="op">)</span></span></code></pre></div>
<p><img src="cdQuPath_files/figure-html/neighborhoodsAna-2.png" width="700"></p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot the clusters for the probability matrix</span></span>
<span><span class="fu"><a href="https://davislaboratory.github.io/hoodscanR/reference/plotTissue.html" class="external-link">plotTissue</a></span><span class="op">(</span><span class="va">spe</span>, color <span class="op">=</span> <span class="va">clusters</span><span class="op">)</span></span></code></pre></div>
<p><img src="cdQuPath_files/figure-html/neighborhoodsAna-3.png" width="700"></p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot heatmap for neighbourhood analysis</span></span>
<span><span class="fu"><a href="https://davislaboratory.github.io/hoodscanR/reference/plotColocal.html" class="external-link">plotColocal</a></span><span class="op">(</span><span class="va">spe</span>, pm_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="cdQuPath_files/figure-html/neighborhoodsAna-4.png" width="700"></p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot probability distribution</span></span>
<span><span class="fu"><a href="https://davislaboratory.github.io/hoodscanR/reference/plotProbDist.html" class="external-link">plotProbDist</a></span><span class="op">(</span><span class="va">spe</span>, pm_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span>, </span>
<span>             by_cluster <span class="op">=</span> <span class="cn">TRUE</span>, plot_all <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>             show_clusters <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="va">clusterK</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="cdQuPath_files/figure-html/neighborhoodsAna-5.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="co-occurrence-analysis">Co-occurrence analysis<a class="anchor" aria-label="anchor" href="#co-occurrence-analysis"></a>
</h4>
<p>The Jaccard Index is a statistic used for gauging the similarity and
diversity of sample sets. The Jaccard Index is defined as:
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mfrac><mrow><mi>c</mi><mi>o</mi><mi>o</mi><mi>c</mi><mi>c</mi><mi>u</mi><mi>r</mi><mi>e</mi><mi>n</mi><mi>t</mi><mi>_</mi><mi>c</mi><mi>e</mi><mi>l</mi><mi>l</mi><mi>t</mi><mi>y</mi><mi>p</mi><mi>e</mi><mi>_</mi><mi>A</mi><mi>_</mi><mi>a</mi><mi>n</mi><mi>d</mi><mi>_</mi><mi>B</mi></mrow><mrow><mi>c</mi><mi>o</mi><mi>u</mi><mi>n</mi><mi>t</mi><mi>_</mi><mi>c</mi><mi>e</mi><mi>l</mi><mi>l</mi><mi>t</mi><mi>y</mi><mi>p</mi><mi>e</mi><mi>_</mi><mi>A</mi><mo>+</mo><mi>c</mi><mi>o</mi><mi>u</mi><mi>n</mi><mi>t</mi><mi>_</mi><mi>c</mi><mi>e</mi><mi>l</mi><mi>l</mi><mi>t</mi><mi>y</mi><mi>p</mi><mi>e</mi><mi>_</mi><mi>B</mi><mo>−</mo><mi>c</mi><mi>o</mi><mi>o</mi><mi>c</mi><mi>c</mi><mi>u</mi><mi>r</mi><mi>e</mi><mi>n</mi><mi>t</mi><mi>_</mi><mi>c</mi><mi>e</mi><mi>l</mi><mi>l</mi><mi>t</mi><mi>y</mi><mi>p</mi><mi>e</mi><mi>_</mi><mi>A</mi><mi>_</mi><mi>a</mi><mi>n</mi><mi>d</mi><mi>_</mi><mi>B</mi></mrow></mfrac><annotation encoding="application/x-tex">\frac{cooccurent\_celltype\_A\_and\_B}{count\_celltype\_A+count\_celltype\_B-cooccurent\_celltype\_A\_and\_B}</annotation></semantics></math></p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">maxgap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">Cell.Length.um</span><span class="op">)</span> <span class="op">*</span> <span class="fl">2</span></span>
<span><span class="va">ji</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/JaccardIndex.html">JaccardIndex</a></span><span class="op">(</span><span class="va">spe</span>, maxgap <span class="op">=</span> <span class="va">maxgap</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotJaccardIndex.html">plotJaccardIndex</a></span><span class="op">(</span><span class="va">ji</span><span class="op">)</span></span></code></pre></div>
<p><img src="cdQuPath_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<p>The co-occurrence score gives us an indication on whether given two
factors such as cell cluster, cell type, and etc, co-occur with each
other at increasing distances across the tissue. The co-occurrence score
is defined as:
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mfrac><mrow><mi>p</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>f</mi><mi>a</mi><mi>c</mi><mi>t</mi><mi>o</mi><mi>r</mi><mn>1</mn><mo stretchy="false" form="prefix">|</mo><mi>f</mi><mi>a</mi><mi>c</mi><mi>t</mi><mi>o</mi><mi>r</mi><mn>2</mn><mo stretchy="true" form="postfix">)</mo></mrow></mrow><mrow><mi>p</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>f</mi><mi>a</mi><mi>c</mi><mi>t</mi><mi>o</mi><mi>r</mi><mn>1</mn><mo stretchy="true" form="postfix">)</mo></mrow></mrow></mfrac><annotation encoding="application/x-tex">\frac{p(factor1|factor2)}{p(factor1)}</annotation></semantics></math></p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## TODO</span></span></code></pre></div>
<p>The spatial score define the distance ratio among three factors. The
spatial score is defined as:
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mfrac><mrow><mi>c</mi><mi>e</mi><mi>l</mi><mi>l</mi><mi>t</mi><mi>y</mi><mi>p</mi><mi>e</mi><mi>C</mi><mi>e</mi><mi>n</mi><mi>t</mi><mi>e</mi><mi>r</mi><mi>_</mi><mi>d</mi><mi>i</mi><mi>s</mi><mi>t</mi><mi>a</mi><mi>n</mi><mi>c</mi><mi>e</mi><mi>_</mi><mi>t</mi><mi>o</mi><mi>_</mi><mi>n</mi><mi>e</mi><mi>a</mi><mi>r</mi><mi>e</mi><mi>s</mi><mi>t</mi><mi>_</mi><mi>c</mi><mi>e</mi><mi>l</mi><mi>l</mi><mi>t</mi><mi>y</mi><mi>p</mi><mi>e</mi><mi>R</mi><mi>i</mi><mi>g</mi><mi>h</mi><mi>t</mi></mrow><mrow><mi>c</mi><mi>e</mi><mi>l</mi><mi>l</mi><mi>t</mi><mi>y</mi><mi>p</mi><mi>e</mi><mi>C</mi><mi>e</mi><mi>n</mi><mi>t</mi><mi>e</mi><mi>r</mi><mi>_</mi><mi>d</mi><mi>i</mi><mi>s</mi><mi>t</mi><mi>a</mi><mi>n</mi><mi>c</mi><mi>e</mi><mi>_</mi><mi>t</mi><mi>o</mi><mi>_</mi><mi>n</mi><mi>e</mi><mi>a</mi><mi>r</mi><mi>e</mi><mi>s</mi><mi>t</mi><mi>_</mi><mi>c</mi><mi>e</mi><mi>l</mi><mi>l</mi><mi>t</mi><mi>y</mi><mi>p</mi><mi>e</mi><mi>L</mi><mi>e</mi><mi>f</mi><mi>t</mi></mrow></mfrac><annotation encoding="application/x-tex">\frac{celltypeCenter\_distance\_to\_nearest\_celltypeRight}{celltypeCenter\_distance\_to\_nearest\_celltypeLeft}</annotation></semantics></math>.</p>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## need to increase the neighbors number</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/neighborhoodsAna.html">neighborhoodsAna</a></span><span class="op">(</span><span class="va">mergedSeu</span>, clusterK <span class="op">=</span> <span class="va">clusterK</span>,</span>
<span>                        k <span class="op">=</span> <span class="fl">500</span>, <span class="co">## maximal 500 neighbors</span></span>
<span>                        anno_col <span class="op">=</span> <span class="st">'celltype_first'</span><span class="op">)</span></span>
<span><span class="va">ss</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spatialScore.html">spatialScore</a></span><span class="op">(</span><span class="va">spe</span>,</span>
<span>                   celltypeCenter <span class="op">=</span> <span class="st">"M2 MACROPHAGE"</span>,</span>
<span>                   celltypeLeft <span class="op">=</span> <span class="st">"Cytotoxic T Cells"</span>,</span>
<span>                   celltypeRight <span class="op">=</span> <span class="st">"Tumor/Myeloid"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ss</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                      spatialScore    distR     distL</span></span>
<span><span class="co">## 8e6b7723-2547-4ad7-8bea-810ff6432fa4    13.374156 353.7328  26.44898</span></span>
<span><span class="co">## aa7a6735-dfb3-480e-a8c1-efc7f45e136b    13.374156 353.7328  26.44898</span></span>
<span><span class="co">## ec87b03d-aa83-4f5d-b9f5-654add3defc7    13.374156 353.7328  26.44898</span></span>
<span><span class="co">## 6aa7e11c-19e1-4a15-b92d-cea75093ffae     2.320486 188.2077  81.10703</span></span>
<span><span class="co">## 22845794-ab72-4356-a185-87d7427e81a7     2.524645 380.9258 150.88294</span></span>
<span><span class="co">## 0944dbdc-b761-4469-bd24-1f8fc1742e50     2.449444 111.8723  45.67253</span></span></code></pre>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html" class="external-link">boxplot</a></span><span class="op">(</span><span class="va">ss</span><span class="op">$</span><span class="va">spatialScore</span><span class="op">)</span></span></code></pre></div>
<p><img src="cdQuPath_files/figure-html/spatialScore-1.png" width="700"></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>R version 4.4.1 (2024-06-14) Platform: x86_64-pc-linux-gnu Running
under: Ubuntu 22.04.4 LTS</p>
<p>Matrix products: default BLAS:
/usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 LAPACK:
/usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;
LAPACK version 3.10.0</p>
<p>locale: [1] LC_CTYPE=en_US.UTF-8 LC_NUMERIC=C<br>
[3] LC_TIME=en_US.UTF-8 LC_COLLATE=en_US.UTF-8<br>
[5] LC_MONETARY=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8<br>
[7] LC_PAPER=en_US.UTF-8 LC_NAME=C<br>
[9] LC_ADDRESS=C LC_TELEPHONE=C<br>
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C</p>
<p>time zone: Etc/UTC tzcode source: system (glibc)</p>
<p>attached base packages: [1] stats4 stats graphics grDevices utils
datasets methods<br>
[8] base</p>
<p>other attached packages: [1] SpatialExperiment_1.15.1
SingleCellExperiment_1.27.2 [3] SummarizedExperiment_1.35.1
Biobase_2.65.0<br>
[5] GenomicRanges_1.57.1 GenomeInfoDb_1.41.1<br>
[7] IRanges_2.39.2 S4Vectors_0.43.2<br>
[9] BiocGenerics_0.51.0 MatrixGenerics_1.17.0<br>
[11] matrixStats_1.3.0 hoodscanR_1.3.3<br>
[13] dittoSeq_1.17.0 ggplot2_3.5.1<br>
[15] Seurat_5.1.0 SeuratObject_5.0.2<br>
[17] sp_2.1-4 cdQuPath_0.99.0.4<br>
[19] BiocStyle_2.33.1</p>
<p>loaded via a namespace (and not attached): [1] RcppAnnoy_0.0.22
splines_4.4.1 later_1.3.2<br>
[4] tibble_3.2.1 polyclip_1.10-7 fastDummies_1.7.4<br>
[7] lifecycle_1.0.4 doParallel_1.0.17 globals_0.16.3<br>
[10] lattice_0.22-6 MASS_7.3-61 magrittr_2.0.3<br>
[13] plotly_4.10.4 sass_0.4.9 rmarkdown_2.28<br>
[16] jquerylib_0.1.4 yaml_2.3.10 httpuv_1.6.15<br>
[19] sctransform_0.4.1 spam_2.10-0 spatstat.sparse_3.1-0<br>
[22] reticulate_1.38.0 cowplot_1.1.3 pbapply_1.7-2<br>
[25] RColorBrewer_1.1-3 abind_1.4-5 zlibbioc_1.51.1<br>
[28] Rtsne_0.17 purrr_1.0.2 circlize_0.4.16<br>
[31] GenomeInfoDbData_1.2.12 ggrepel_0.9.5 irlba_2.3.5.1<br>
[34] listenv_0.9.1 spatstat.utils_3.1-0 pheatmap_1.0.12<br>
[37] goftest_1.2-3 RSpectra_0.16-2 spatstat.random_3.3-1<br>
[40] fitdistrplus_1.2-1 parallelly_1.38.0 pkgdown_2.1.0<br>
[43] leiden_0.4.3.1 codetools_0.2-20 DelayedArray_0.31.11<br>
[46] shape_1.4.6.1 tidyselect_1.2.1 UCSC.utils_1.1.0<br>
[49] farver_2.1.2 spatstat.explore_3.3-1 jsonlite_1.8.8<br>
[52] GetoptLong_1.0.5 e1071_1.7-14 progressr_0.14.0<br>
[55] iterators_1.0.14 ggridges_0.5.6 survival_3.7-0<br>
[58] systemfonts_1.1.0 foreach_1.5.2 tools_4.4.1<br>
[61] ragg_1.3.2 ica_1.0-3 Rcpp_1.0.13<br>
[64] glue_1.7.0 gridExtra_2.3 SparseArray_1.5.31<br>
[67] xfun_0.47 dplyr_1.1.4 withr_3.0.1<br>
[70] BiocManager_1.30.23 fastmap_1.2.0 fansi_1.0.6<br>
[73] digest_0.6.37 R6_2.5.1 mime_0.12<br>
[76] textshaping_0.4.0 colorspace_2.1-1 scattermore_1.2<br>
[79] tensor_1.5 spatstat.data_3.1-2 utf8_1.2.4<br>
[82] tidyr_1.3.1 generics_0.1.3 data.table_1.15.4<br>
[85] class_7.3-22 httr_1.4.7 htmlwidgets_1.6.4<br>
[88] S4Arrays_1.5.7 uwot_0.2.2 pkgconfig_2.0.3<br>
[91] scico_1.5.0 gtable_0.3.5 ComplexHeatmap_2.21.0<br>
[94] lmtest_0.9-40 XVector_0.45.0 htmltools_0.5.8.1<br>
[97] dotCall64_1.1-1 bookdown_0.40 clue_0.3-65<br>
[100] scales_1.3.0 png_0.1-8 spatstat.univar_3.0-0<br>
[103] knitr_1.48 reshape2_1.4.4 rjson_0.2.22<br>
[106] nlme_3.1-166 GlobalOptions_0.1.2 proxy_0.4-27<br>
[109] cachem_1.1.0 zoo_1.8-12 stringr_1.5.1<br>
[112] KernSmooth_2.23-24 parallel_4.4.1 miniUI_0.1.1.1<br>
[115] desc_1.4.3 pillar_1.9.0 grid_4.4.1<br>
[118] vctrs_0.6.5 RANN_2.6.1 promises_1.3.0<br>
[121] xtable_1.8-4 cluster_2.1.6 evaluate_0.24.0<br>
[124] magick_2.8.4 cli_3.6.3 compiler_4.4.1<br>
[127] rlang_1.1.4 crayon_1.5.3 future.apply_1.11.2<br>
[130] labeling_0.4.3 mclust_6.1.1 plyr_1.8.9<br>
[133] fs_1.6.4 stringi_1.8.4 viridisLite_0.4.2<br>
[136] deldir_2.0-4 munsell_0.5.1 lazyeval_0.2.2<br>
[139] spatstat.geom_3.3-2 Matrix_1.7-0 RcppHNSW_0.6.0<br>
[142] patchwork_1.2.0 future_1.34.0 shiny_1.9.1<br>
[145] highr_0.11 ROCR_1.0-11 igraph_2.0.3<br>
[148] bslib_0.8.0</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jianhong Ou.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
